### ISSUE TITLE
Gateway API HttpRoute URLRewrite not working

### ISSUE THREAD

**Issue by nxy7** [cilium/cilium#27954]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

After having few issues with setting up cilium (#27723) I've finally got to the root issue - missing kernel modules. I've set up my system correctly now, and can access HttpRoutes through gateway, but some features like URLRewrite are not working (I've confirmed it with echoserver). I've checked logs of "Cilium", "Cilium Operator" and "Cilium Envoy", but there are no suspicious warnings of any kind.
There are some issues in cilium sysdump, but they're vague and hard to decipher without knowing cilium internals.

### Cilium Version

1.14.1

### Kernel Version

6.5

### Kubernetes Version

Client Version: v1.28.1
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.27.4+k0s

### Sysdump

[cilium-sysdump-20230906-003232.zip](https://github.com/cilium/cilium/files/12528913/cilium-sysdump-20230906-003232.zip)

### Relevant log output

```shell
CILIUM SYSDUMP Suspicious warnings

 Collecting gops stats from Hubble pods
level=warning msg="Ignoring invalid CiliumEnvoyConfig JSON (proto: (line 1:405): unable to resolve \"type.googleapis.com/envoy.extensions.filters.http.router.v3.Router\": \"not found\"): {\n\t\"@type\": \"type.googleapis.com/envoy.config.listener.v3.Listener\",\n\t\"filterChains\": [\n\t\t{\n\t\t\t\"filterChainMatch\": {\n\t\t\t\t\"transportProtocol\": \"raw_buffer\"\n\t\t\t},\n\t\t\t\"filters\": [\n\t\t\t\t{\n\t\t\t\t\t\"name\": \"envoy.filters.network.http_connection_manager\",\n\t\t\t\t\t\"typedConfig\": {\n\t\t\t\t\t\t\"@type\": \"type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager\",\n\t\t\t\t\t\t\"httpFilters\": [\n\t\t\t\t\t\t\t{\n\t\t\t\t\t\t\t\t\"name\": \"envoy.filters.http.router\",\n\t\t\t\t\t\t\t\t\"typedConfig\": {\n\t\t\t\t\t\t\t\t\t\"@type\": \"type.googleapis.com/envoy.extensions.filters.http.router.v3.Router\"\n\t\t\t\t\t\t\t\t}\n\t\t\t\t\t\t\t}\n\t\t\t\t\t\t],\n\t\t\t\t\t\t\"rds\": {\n\t\t\t\t\t\t\t\"routeConfigName\": \"listener-insecure\"\n\t\t\t\t\t\t},\n\t\t\t\t\t\t\"statPrefix\": \"listener-insecure\",\n\t\t\t\t\t\t\"upgradeConfigs\": [\n\t\t\t\t\t\t\t{\n\t\t\t\t\t\t\t\t\"upgradeType\": \"websocket\"\n\t\t\t\t\t\t\t}\n\t\t\t\t\t\t],\n\t\t\t\t\t\t\"useRemoteAddress\": true\n\t\t\t\t\t}\n\t\t\t\t}\n\t\t\t]\n\t\t}\n\t],\n\t\"listenerFilters\": [\n\t\t{\n\t\t\t\"name\": \"envoy.filters.listener.tls_inspector\",\n\t\t\t\"typedConfig\": {\n\t\t\t\t\"@type\": \"type.googleapis.com/envoy.extensions.filters.listener.tls_inspector.v3.TlsInspector\"\n\t\t\t}\n\t\t}\n\t],\n\t\"name\": \"listener\",\n\t\"socketOptions\": [\n\t\t{\n\t\t\t\"description\": \"Enable TCP keep-alive (default to enabled)\",\n\t\t\t\"intValue\": \"1\",\n\t\t\t\"level\": \"1\",\n\t\t\t\"name\": \"9\",\n\t\t\t\"state\": \"STATE_LISTENING\"\n\t\t},\n\t\t{\n\t\t\t\"description\": \"TCP keep-alive idle time (in seconds) (defaults to 10s)\",\n\t\t\t\"intValue\": \"10\",\n\t\t\t\"level\": \"6\",\n\t\t\t\"name\": \"4\",\n\t\t\t\"state\": \"STATE_LISTENING\"\n\t\t},\n\t\t{\n\t\t\t\"description\": \"TCP keep-alive probe intervals (in seconds) (defaults to 5s)\",\n\t\t\t\"intValue\": \"5\",\n\t\t\t\"level\": \"6\",\n\t\t\t\"name\": \"5\",\n\t\t\t\"state\": \"STATE_LISTENING\"\n\t\t},\n\t\t{\n\t\t\t\"description\": \"TCP keep-alive probe max failures.\",\n\t\t\t\"intValue\": \"10\",\n\t\t\t\"level\": \"6\",\n\t\t\t\"name\": \"6\",\n\t\t\t\"state\": \"STATE_LISTENING\"\n\t\t}\n\t]\n}" subsys=k8s-apis-cilium.io-v2
⚠️ Daemonset "cilium-node-init" not found in namespace "kube-system" - this is expected if Node Init DaemonSet is not enabled
🔍 Collecting gops stats from Hubble Relay pods
level=warning msg="Ignoring invalid CiliumEnvoyConfig JSON (proto: (line 1:289): unable to resolve \"type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions\": \"not found\"): {\n\t\"@type\": \"type.googleapis.com/envoy.config.cluster.v3.Cluster\",\n\t\"connectTimeout\": \"5s\",\n\t\"name\": \"streampai/backend:7000\",\n\t\"outlierDetection\": {\n\t\t\"splitExternalLocalOriginErrors\": true\n\t},\n\t\"type\": \"EDS\",\n\t\"typedExtensionProtocolOptions\": {\n\t\t\"envoy.extensions.upstreams.http.v3.HttpProtocolOptions\": {\n\t\t\t\"@type\": \"type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions\",\n\t\t\t\"commonHttpProtocolOptions\": {\n\t\t\t\t\"idleTimeout\": \"60s\"\n\t\t\t},\n\t\t\t\"useDownstreamProtocolConfig\": {\n\t\t\t\t\"http2ProtocolOptions\": {}\n\t\t\t}\n\t\t}\n\t}\n}" subsys=k8s-apis-cilium.io-v2
level=warning msg="Ignoring invalid CiliumEnvoyConfig JSON (proto: (line 1:290): unable to resolve \"type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions\": \"not found\"): {\n\t\"@type\": \"type.googleapis.com/envoy.config.cluster.v3.Cluster\",\n\t\"connectTimeout\": \"5s\",\n\t\"name\": \"streampai/frontend:3000\",\n\t\"outlierDetection\": {\n\t\t\"splitExternalLocalOriginErrors\": true\n\t},\n\t\"type\": \"EDS\",\n\t\"typedExtensionProtocolOptions\": {\n\t\t\"envoy.extensions.upstreams.http.v3.HttpProtocolOptions\": {\n\t\t\t\"@type\": \"type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions\",\n\t\t\t\"commonHttpProtocolOptions\": {\n\t\t\t\t\"idleTimeout\": \"60s\"\n\t\t\t},\n\t\t\t\"useDownstreamProtocolConfig\": {\n\t\t\t\t\"http2ProtocolOptions\": {}\n\t\t\t}\n\t\t}\n\t}\n}" subsys=k8s-apis-cilium.io-v2
level=warning msg="Ignoring invalid CiliumEnvoyConfig JSON (proto: (line 1:288): unable to resolve \"type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions\": \"not found\"): {\n\t\"@type\": \"type.googleapis.com/envoy.config.cluster.v3.Cluster\",\n\t\"connectTimeout\": \"5s\",\n\t\"name\": \"streampai/kratos:4433\",\n\t\"outlierDetection\": {\n\t\t\"splitExternalLocalOriginErrors\": true\n\t},\n\t\"type\": \"EDS\",\n\t\"typedExtensionProtocolOptions\": {\n\t\t\"envoy.extensions.upstreams.http.v3.HttpProtocolOptions\": {\n\t\t\t\"@type\": \"type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions\",\n\t\t\t\"commonHttpProtocolOptions\": {\n\t\t\t\t\"idleTimeout\": \"60s\"\n\t\t\t},\n\t\t\t\"useDownstreamProtocolConfig\": {\n\t\t\t\t\"http2ProtocolOptions\": {}\n\t\t\t}\n\t\t}\n\t}\n}" subsys=k8s-apis-cilium.io-v2
level=warning msg="Ignoring invalid CiliumEnvoyConfig JSON (proto: (line 1:287): unable to resolve \"type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions\": \"not found\"): {\n\t\"@type\": \"type.googleapis.com/envoy.config.cluster.v3.Cluster\",\n\t\"connectTimeout\": \"5s\",\n\t\"name\": \"streampai/minio:9000\",\n\t\"outlierDetection\": {\n\t\t\"splitExternalLocalOriginErrors\": true\n\t},\n\t\"type\": \"EDS\",\n\t\"typedExtensionProtocolOptions\": {\n\t\t\"envoy.extensions.upstreams.http.v3.HttpProtocolOptions\": {\n\t\t\t\"@type\": \"type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions\",\n\t\t\t\"commonHttpProtocolOptions\": {\n\t\t\t\t\"idleTimeout\": \"60s\"\n\t\t\t},\n\t\t\t\"useDownstreamProtocolConfig\": {\n\t\t\t\t\"http2ProtocolOptions\": {}\n\t\t\t}\n\t\t}\n\t}\n}" subsys=k8s-apis-cilium.io-v2
level=warning msg="Ignoring invalid CiliumEnvoyConfig JSON (proto: (line 1:287): unable to resolve \"type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions\": \"not found\"): {\n\t\"@type\": \"type.googleapis.com/envoy.config.cluster.v3.Cluster\",\n\t\"connectTimeout\": \"5s\",\n\t\"name\": \"streampai/pgweb:8081\",\n\t\"outlierDetection\": {\n\t\t\"splitExternalLocalOriginErrors\": true\n\t},\n\t\"type\": \"EDS\",\n\t\"typedExtensionProtocolOptions\": {\n\t\t\"envoy.extensions.upstreams.http.v3.HttpProtocolOptions\": {\n\t\t\t\"@type\": \"type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions\",\n\t\t\t\"commonHttpProtocolOptions\": {\n\t\t\t\t\"idleTimeout\": \"60s\"\n\t\t\t},\n\t\t\t\"useDownstreamProtocolConfig\": {\n\t\t\t\t\"http2ProtocolOptions\": {}\n\t\t\t}\n\t\t}\n\t}\n}" subsys=k8s-apis-cilium.io-v2
```

### Anything else?

There's one more thing that's weird within logs. 
```
[2023-09-05 22:29:08.728][1][debug][connection] [external/envoy/source/common/network/connection_impl.cc:939] [C1] connecting to /var/run/cilium/envoy/sockets/xds.sock
[2023-09-05 22:29:08.728][1][debug][connection] [external/envoy/source/common/network/connection_impl.cc:963] [C1] immediate connect error: No such file or directory
```
I've checked and /var/run/cilium/envoy/sockets/xds.sock is not missing, but this second log could actually be refering to some other file.

Given that most of my previous issues were linked to missing kernel modules, does URLRewrite need any kernel modules? If not what might be the reason that it's not working and what would be the steps to debug it?

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by youngnick** [comment]
Thanks for this issue @nxy7 - we've only added support for `urlRewrite` and some other extended features in `main` at the moment, that will be included in Cilium 1.15. So that's why it's not working.

**Comment by nxy7** [comment]
Oh, for some reason I was sure it worked some time ago when I checked out cilium. Sorry for opening this issue then :P I don't think it's available on 1.15.0-pre.0 so I'll be waiting until 1.15.0 is released. Thanks for answer.
